File: python_soluciones/challenge_01.py
"""
    Reto #1
    ¿ES UN ANAGRAMA?
    Fecha publicación enunciado: 2022/01/
    Fecha publicación resolución: 2023/04/18
    Dificultad: MEDIA

    Enunciado: Escribe una función que reciba dos palabras (String)
        y retorne verdadero o falso (Boolean) según sean o no anagramas.
    Un Anagrama consiste en formar una palabra reordenando TODAS las letras de otra palabra inicial.
    NO hace falta comprobar que ambas palabras existan.
    Dos palabras exactamente iguales no son anagrama.
"""


def main():
    """Función principal"""

    texto_01 = input("Escriba una palabra: ")
    texto_02 = input("Escriba una segunda palabra: ")

    son_anagramas = anagrama(texto_01, texto_02)
    if son_anagramas:
        print("Las palabras SON ANAGRAMAS")
    else:
        print("Las palabras NO son ANAGRAMAS")


def anagrama(texto_01: str, texto_02: str) -> bool:
    """Recibe dos textos y determina si es un anagrama"""

    # Como lo aclara el reto, dos textos iguales no son un anagrama
    if texto_01 == texto_02:
        return False

    # Si tiene diferente longitud lo más seguro es que sean diferentes
    if len(texto_01) != len(texto_02):
        return False

    # Revisar el texto_01 contiene cada una de sus letras
    for letra in texto_01:
        if not letra in texto_02:
            return False
    # Comprobamos en el otro sentido, si el texto_02 contiene todas las letras del texto_01
    for letra in texto_02:
        if not letra in texto_01:
            return False

    # Entrega True si pasó todas las pruebas
    return True


if __name__ == "__main__":
    main()

File: app/src/main/java/com/nadisla/weeklychallenge2022/challenge01-dart/challenge01.dart
/*
* Challenge #1
* IS IT AN ANAGRAM?
* Difficulty: MEDIUM
*
* Statement: Write a function that receives two words (String) and returns true or false (Boolean) depending on whether or not they are anagrams.
* An Anagram consists of forming a word by rearranging ALL the letters of another initial word.
* It is NOT necessary to check that both words exist.
* Two exactly the same words are not anagram.

* Reto #1
* ¿ES UN ANAGRAMA?
* Dificultad: MEDIA
*
* Enunciado: Escribe una función que reciba dos palabras (String) y retorne verdadero o falso (Boolean) según sean o no anagramas.
* Un Anagrama consiste en formar una palabra reordenando TODAS las letras de otra palabra inicial.
* NO hace falta comprobar que ambas palabras existan.
* Dos palabras exactamente iguales no son anagrama.
*/
import 'dart:io';

void main() {
  stdout.write("Enter the first word: ");
  String word1 = stdin.readLineSync()!;
  stdout.write("Enter the second word: ");
  String word2 = stdin.readLineSync()!;
  print("Is the second word an anagram? ${isAnagram(word1, word2)}");
}

bool isAnagram(String word1, String word2) {
  List<String> word1Splitted = word1.toLowerCase().trim().split(""),
      word2Splitted = word2.toLowerCase().trim().split("");
  word1Splitted.sort();
  word2Splitted.sort();

  if (word1 == word2 ||
      word1Splitted.length != word2Splitted.length ||
      word1Splitted.join() != word2Splitted.join()) {
    return false;
  }
  return true;
}
